// typescript-to-lua/dist/transformation/visitors/call.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformSpreadElement = exports.transformCallExpression = exports.transformContextualCallExpression = exports.transformArguments = exports.flattenSpreadExpressions = void 0;
const ts = require("typescript");
const lua = require("../../LuaAST");
const builtins_1 = require("../builtins");
const annotations_1 = require("../utils/annotations");
const assignment_validation_1 = require("../utils/assignment-validation");
const function_context_1 = require("../utils/function-context");
const lua_ast_1 = require("../utils/lua-ast");
const lualib_1 = require("../utils/lualib");
const safe_names_1 = require("../utils/safe-names");
const typescript_1 = require("../utils/typescript");
const access_1 = require("./access");
const lua_table_1 = require("./lua-table");
const multi_1 = require("./language-extensions/multi");
function getExpressionsBeforeAndAfterFirstSpread(expressions) {
    // [a, b, ...c, d, ...e] --> [a, b] and [...c, d, ...e]
    const index = expressions.findIndex(ts.isSpreadElement);
    const hasSpreadElement = index !== -1;
    const before = hasSpreadElement ? expressions.slice(0, index) : expressions;
    const after = hasSpreadElement ? expressions.slice(index) : [];
    return [before, after];
}
function transformSpreadableExpressionsIntoArrayConcatArguments(context, expressions) {
    // [...array, a, b, ...tuple()] --> [ [...array], [a, b], [...tuple()] ]
    // chunk non-spread arguments together so they don't concat
    const chunks = [];
    for (const [index, expression] of expressions.entries()) {
        if (ts.isSpreadElement(expression)) {
            chunks.push([expression]);
            const next = expressions[index + 1];
            if (next && !ts.isSpreadElement(next)) {
                chunks.push([]);
            }
        }
        else {
            let lastChunk = chunks[chunks.length - 1];
            if (!lastChunk) {
                lastChunk = [];
                chunks.push(lastChunk);
            }
            lastChunk.push(expression);
        }
    }
    return chunks.map(chunk => lua_ast_1.wrapInTable(...chunk.map(expression => context.transformExpression(expression))));
}
function flattenSpreadExpressions(context, expressions) {
    const [preSpreadExpressions, postSpreadExpressions] = getExpressionsBeforeAndAfterFirstSpread(expressions);
    const transformedPreSpreadExpressions = preSpreadExpressions.map(a => context.transformExpression(a));
    // Nothing special required
    if (postSpreadExpressions.length === 0) {
        return transformedPreSpreadExpressions;
    }
    // Only one spread element at the end? Will work as expected
    if (postSpreadExpressions.length === 1) {
        return [...transformedPreSpreadExpressions, context.transformExpression(postSpreadExpressions[0])];
    }
    // Use Array.concat and unpack the result of that as the last Expression
    const concatArguments = transformSpreadableExpressionsIntoArrayConcatArguments(context, postSpreadExpressions);
    const lastExpression = lua_ast_1.createUnpackCall(context, lualib_1.transformLuaLibFunction(context, lualib_1.LuaLibFeature.ArrayConcat, undefined, ...concatArguments));
    return [...transformedPreSpreadExpressions, lastExpression];
}
exports.flattenSpreadExpressions = flattenSpreadExpressions;
function transformArguments(context, params, signature, callContext) {
    const parameters = flattenSpreadExpressions(context, params);
    // Add context as first param if present
    if (callContext) {
        parameters.unshift(context.transformExpression(callContext));
    }
    if (signature && signature.parameters.length >= params.length) {
        for (const [index, param] of params.entries()) {
            const signatureParameter = signature.parameters[index];
            const paramType = context.checker.getTypeAtLocation(param);
            const signatureType = context.checker.getTypeAtLocation(signatureParameter.valueDeclaration);
            assignment_validation_1.validateAssignment(context, param, paramType, signatureType, signatureParameter.name);
        }
    }
    return parameters;
}
exports.transformArguments = transformArguments;
function transformContextualCallExpression(context, node, transformedArguments) {
    const left = ts.isCallExpression(node) ? node.expression : node.tag;
    if (ts.isPropertyAccessExpression(left) && ts.isIdentifier(left.name) && safe_names_1.isValidLuaIdentifier(left.name.text)) {
        // table:name()
        const table = context.transformExpression(left.expression);
        return lua.createMethodCallExpression(table, lua.createIdentifier(left.name.text, left.name), transformedArguments, node);
    }
    else if (ts.isElementAccessExpression(left) || ts.isPropertyAccessExpression(left)) {
        const callContext = context.transformExpression(left.expression);
        if (typescript_1.isExpressionWithEvaluationEffect(left.expression)) {
            // Inject context parameter
            transformedArguments.unshift(lua.createIdentifier("____self"));
            // Cache left-side if it has effects
            // (function() local ____self = context; return ____self[argument](parameters); end)()
            const argument = ts.isElementAccessExpression(left)
                ? access_1.transformElementAccessArgument(context, left)
                : lua.createStringLiteral(left.name.text);
            const selfIdentifier = lua.createIdentifier("____self");
            const selfAssignment = lua.createVariableDeclarationStatement(selfIdentifier, callContext);
            const index = lua.createTableIndexExpression(selfIdentifier, argument);
            const callExpression = lua.createCallExpression(index, transformedArguments);
            return lua_ast_1.createImmediatelyInvokedFunctionExpression([selfAssignment], callExpression, node);
        }
        else {
            const expression = context.transformExpression(left);
            return lua.createCallExpression(expression, [callContext, ...transformedArguments]);
        }
    }
    else if (ts.isIdentifier(left)) {
        const callContext = context.isStrict ? lua.createNilLiteral() : lua.createIdentifier("_G");
        transformedArguments.unshift(callContext);
        const expression = context.transformExpression(left);
        return lua.createCallExpression(expression, transformedArguments, node);
    }
    else {
        throw new Error(`Unsupported LeftHandSideExpression kind: ${ts.SyntaxKind[left.kind]}`);
    }
}
exports.transformContextualCallExpression = transformContextualCallExpression;
function transformPropertyCall(context, node) {
    const signature = context.checker.getResolvedSignature(node);
    if (node.expression.expression.kind === ts.SyntaxKind.SuperKeyword) {
        // Super calls take the format of super.call(self,...)
        const parameters = transformArguments(context, node.arguments, signature, ts.createThis());
        return lua.createCallExpression(context.transformExpression(node.expression), parameters);
    }
    const parameters = transformArguments(context, node.arguments, signature);
    const signatureDeclaration = signature === null || signature === void 0 ? void 0 : signature.getDeclaration();
    if (!signatureDeclaration || function_context_1.getDeclarationContextType(context, signatureDeclaration) !== function_context_1.ContextType.Void) {
        // table:name()
        return transformContextualCallExpression(context, node, parameters);
    }
    else {
        const table = context.transformExpression(node.expression.expression);
        // table.name()
        const name = node.expression.name.text;
        const callPath = lua.createTableIndexExpression(table, lua.createStringLiteral(name), node.expression);
        return lua.createCallExpression(callPath, parameters, node);
    }
}
function transformElementCall(context, node) {
    const signature = context.checker.getResolvedSignature(node);
    const signatureDeclaration = signature === null || signature === void 0 ? void 0 : signature.getDeclaration();
    const parameters = transformArguments(context, node.arguments, signature);
    if (!signatureDeclaration || function_context_1.getDeclarationContextType(context, signatureDeclaration) !== function_context_1.ContextType.Void) {
        // A contextual parameter must be given to this call expression
        return transformContextualCallExpression(context, node, parameters);
    }
    else {
        // No context
        const expression = context.transformExpression(node.expression);
        return lua.createCallExpression(expression, parameters);
    }
}
exports.transformCallExpression = (node, context) => {
    const luaTableResult = lua_table_1.transformLuaTableCallExpression(context, node);
    if (luaTableResult) {
        return luaTableResult;
    }
    const isTupleReturn = annotations_1.isTupleReturnCall(context, node);
    const isTupleReturnForward = node.parent && ts.isReturnStatement(node.parent) && annotations_1.isInTupleReturnFunction(context, node);
    const isInSpread = node.parent && ts.isSpreadElement(node.parent);
    const returnValueIsUsed = node.parent && !ts.isExpressionStatement(node.parent);
    const wrapResult = isTupleReturn && !isTupleReturnForward && !typescript_1.isInDestructingAssignment(node) && !isInSpread && returnValueIsUsed;
    const builtinResult = builtins_1.transformBuiltinCallExpression(context, node);
    if (builtinResult) {
        return wrapResult ? lua_ast_1.wrapInTable(builtinResult) : builtinResult;
    }
    if (ts.isPropertyAccessExpression(node.expression)) {
        const result = transformPropertyCall(context, node);
        return wrapResult ? lua_ast_1.wrapInTable(result) : result;
    }
    if (ts.isElementAccessExpression(node.expression)) {
        const result = transformElementCall(context, node);
        return wrapResult ? lua_ast_1.wrapInTable(result) : result;
    }
    const signature = context.checker.getResolvedSignature(node);
    // Handle super calls properly
    if (node.expression.kind === ts.SyntaxKind.SuperKeyword) {
        const parameters = transformArguments(context, node.arguments, signature, ts.createThis());
        return lua.createCallExpression(lua.createTableIndexExpression(context.transformExpression(ts.createSuper()), lua.createStringLiteral("____constructor")), parameters);
    }
    const callPath = context.transformExpression(node.expression);
    const signatureDeclaration = signature === null || signature === void 0 ? void 0 : signature.getDeclaration();
    let parameters = [];
    if (signatureDeclaration && function_context_1.getDeclarationContextType(context, signatureDeclaration) === function_context_1.ContextType.Void) {
        parameters = transformArguments(context, node.arguments, signature);
    }
    else {
        const callContext = context.isStrict ? ts.createNull() : ts.createIdentifier("_G");
        parameters = transformArguments(context, node.arguments, signature, callContext);
    }
    const callExpression = lua.createCallExpression(callPath, parameters, node);
    return wrapResult ? lua_ast_1.wrapInTable(callExpression) : callExpression;
};
// TODO: Currently it's also used as an array member
exports.transformSpreadElement = (node, context) => {
    const innerExpression = context.transformExpression(node.expression);
    if (annotations_1.isTupleReturnCall(context, node.expression))
        return innerExpression;
    if (ts.isCallExpression(node.expression) && multi_1.returnsMultiType(context, node.expression))
        return innerExpression;
    if (ts.isIdentifier(node.expression) && annotations_1.isVarargType(context, node.expression)) {
        return lua.createDotsLiteral(node);
    }
    const type = context.checker.getTypeAtLocation(node.expression);
    if (typescript_1.isArrayType(context, type)) {
        return lua_ast_1.createUnpackCall(context, innerExpression, node);
    }
    return lualib_1.transformLuaLibFunction(context, lualib_1.LuaLibFeature.Spread, node, innerExpression);
};
