// typescript-to-lua/dist/CompilerOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateOptions = exports.isBundleEnabled = exports.LuaTarget = exports.LuaLibImportKind = void 0;
const diagnosticFactories = require("./transpilation/diagnostics");
var LuaLibImportKind;
(function (LuaLibImportKind) {
    LuaLibImportKind["None"] = "none";
    LuaLibImportKind["Always"] = "always";
    LuaLibImportKind["Inline"] = "inline";
    LuaLibImportKind["Require"] = "require";
})(LuaLibImportKind = exports.LuaLibImportKind || (exports.LuaLibImportKind = {}));
var LuaTarget;
(function (LuaTarget) {
    LuaTarget["Universal"] = "universal";
    LuaTarget["Lua51"] = "5.1";
    LuaTarget["Lua52"] = "5.2";
    LuaTarget["Lua53"] = "5.3";
    LuaTarget["LuaJIT"] = "JIT";
})(LuaTarget = exports.LuaTarget || (exports.LuaTarget = {}));
exports.isBundleEnabled = (options) => options.luaBundle !== undefined && options.luaBundleEntry !== undefined;
function validateOptions(options) {
    const diagnostics = [];
    if (options.luaBundle && !options.luaBundleEntry) {
        diagnostics.push(diagnosticFactories.luaBundleEntryIsRequired());
    }
    if (options.luaBundle && options.luaLibImport === LuaLibImportKind.Inline) {
        diagnostics.push(diagnosticFactories.usingLuaBundleWithInlineMightGenerateDuplicateCode());
    }
    return diagnostics;
}
exports.validateOptions = validateOptions;
