// typescript-to-lua/dist/LuaAST.js
"use strict";
// Simplified Lua AST based roughly on http://lua-users.org/wiki/MetaLuaAbstractSyntaxTree,
// https://www.lua.org/manual/5.3/manual.html#9 and the TS AST implementation
Object.defineProperty(exports, "__esModule", { value: true });
exports.isStringLiteral = exports.createNumericLiteral = exports.isNumericLiteral = exports.createArgLiteral = exports.isArgLiteral = exports.createDotsLiteral = exports.isDotsLiteral = exports.createBooleanLiteral = exports.isBooleanLiteral = exports.createNilLiteral = exports.isNilLiteral = exports.createExpressionStatement = exports.isExpressionStatement = exports.createBreakStatement = exports.isBreakStatement = exports.createReturnStatement = exports.isReturnStatement = exports.createLabelStatement = exports.isLabelStatement = exports.createGotoStatement = exports.isGotoStatement = exports.createForInStatement = exports.isForInStatement = exports.createForStatement = exports.isForStatement = exports.createRepeatStatement = exports.isRepeatStatement = exports.createWhileStatement = exports.isWhileStatement = exports.isIterationStatement = exports.createIfStatement = exports.isIfStatement = exports.createAssignmentStatement = exports.isAssignmentStatement = exports.createVariableDeclarationStatement = exports.isVariableDeclarationStatement = exports.createDoStatement = exports.isDoStatement = exports.createBlock = exports.isBlock = exports.createFile = exports.isFile = exports.setNodeFlags = exports.getOriginalPos = exports.setNodeOriginal = exports.setNodePosition = exports.cloneNode = exports.createNode = exports.NodeFlags = exports.SyntaxKind = void 0;
exports.isInlineFunctionExpression = exports.isFunctionDefinition = exports.isAssignmentLeftHandSideExpression = exports.createTableIndexExpression = exports.isTableIndexExpression = exports.createAnonymousIdentifier = exports.cloneIdentifier = exports.createIdentifier = exports.isIdentifier = exports.createMethodCallExpression = exports.isMethodCallExpression = exports.createCallExpression = exports.isCallExpression = exports.createBinaryExpression = exports.isBinaryExpression = exports.createUnaryExpression = exports.isUnaryExpression = exports.createTableExpression = exports.isTableExpression = exports.createTableFieldExpression = exports.isTableFieldExpression = exports.createFunctionExpression = exports.isFunctionExpression = exports.isLiteral = exports.createStringLiteral = void 0;
// We can elide a lot of nodes especially tokens and keywords
// because we don't create the AST from text
const ts = require("typescript");
const utils_1 = require("./utils");
var SyntaxKind;
(function (SyntaxKind) {
    SyntaxKind[SyntaxKind["File"] = 0] = "File";
    SyntaxKind[SyntaxKind["Block"] = 1] = "Block";
    // Statements
    SyntaxKind[SyntaxKind["DoStatement"] = 2] = "DoStatement";
    SyntaxKind[SyntaxKind["VariableDeclarationStatement"] = 3] = "VariableDeclarationStatement";
    SyntaxKind[SyntaxKind["AssignmentStatement"] = 4] = "AssignmentStatement";
    SyntaxKind[SyntaxKind["IfStatement"] = 5] = "IfStatement";
    SyntaxKind[SyntaxKind["WhileStatement"] = 6] = "WhileStatement";
    SyntaxKind[SyntaxKind["RepeatStatement"] = 7] = "RepeatStatement";
    SyntaxKind[SyntaxKind["ForStatement"] = 8] = "ForStatement";
    SyntaxKind[SyntaxKind["ForInStatement"] = 9] = "ForInStatement";
    SyntaxKind[SyntaxKind["GotoStatement"] = 10] = "GotoStatement";
    SyntaxKind[SyntaxKind["LabelStatement"] = 11] = "LabelStatement";
    SyntaxKind[SyntaxKind["ReturnStatement"] = 12] = "ReturnStatement";
    SyntaxKind[SyntaxKind["BreakStatement"] = 13] = "BreakStatement";
    SyntaxKind[SyntaxKind["ExpressionStatement"] = 14] = "ExpressionStatement";
    // Expression
    SyntaxKind[SyntaxKind["StringLiteral"] = 15] = "StringLiteral";
    SyntaxKind[SyntaxKind["NumericLiteral"] = 16] = "NumericLiteral";
    SyntaxKind[SyntaxKind["NilKeyword"] = 17] = "NilKeyword";
    SyntaxKind[SyntaxKind["DotsKeyword"] = 18] = "DotsKeyword";
    SyntaxKind[SyntaxKind["ArgKeyword"] = 19] = "ArgKeyword";
    SyntaxKind[SyntaxKind["TrueKeyword"] = 20] = "TrueKeyword";
    SyntaxKind[SyntaxKind["FalseKeyword"] = 21] = "FalseKeyword";
    SyntaxKind[SyntaxKind["FunctionExpression"] = 22] = "FunctionExpression";
    SyntaxKind[SyntaxKind["TableFieldExpression"] = 23] = "TableFieldExpression";
    SyntaxKind[SyntaxKind["TableExpression"] = 24] = "TableExpression";
    SyntaxKind[SyntaxKind["UnaryExpression"] = 25] = "UnaryExpression";
    SyntaxKind[SyntaxKind["BinaryExpression"] = 26] = "BinaryExpression";
    SyntaxKind[SyntaxKind["CallExpression"] = 27] = "CallExpression";
    SyntaxKind[SyntaxKind["MethodCallExpression"] = 28] = "MethodCallExpression";
    SyntaxKind[SyntaxKind["Identifier"] = 29] = "Identifier";
    SyntaxKind[SyntaxKind["TableIndexExpression"] = 30] = "TableIndexExpression";
    // Operators
    // Arithmetic
    SyntaxKind[SyntaxKind["AdditionOperator"] = 31] = "AdditionOperator";
    SyntaxKind[SyntaxKind["SubtractionOperator"] = 32] = "SubtractionOperator";
    SyntaxKind[SyntaxKind["MultiplicationOperator"] = 33] = "MultiplicationOperator";
    SyntaxKind[SyntaxKind["DivisionOperator"] = 34] = "DivisionOperator";
    SyntaxKind[SyntaxKind["FloorDivisionOperator"] = 35] = "FloorDivisionOperator";
    SyntaxKind[SyntaxKind["ModuloOperator"] = 36] = "ModuloOperator";
    SyntaxKind[SyntaxKind["PowerOperator"] = 37] = "PowerOperator";
    SyntaxKind[SyntaxKind["NegationOperator"] = 38] = "NegationOperator";
    // Concat
    SyntaxKind[SyntaxKind["ConcatOperator"] = 39] = "ConcatOperator";
    // Length
    SyntaxKind[SyntaxKind["LengthOperator"] = 40] = "LengthOperator";
    // Relational Ops
    SyntaxKind[SyntaxKind["EqualityOperator"] = 41] = "EqualityOperator";
    SyntaxKind[SyntaxKind["InequalityOperator"] = 42] = "InequalityOperator";
    SyntaxKind[SyntaxKind["LessThanOperator"] = 43] = "LessThanOperator";
    SyntaxKind[SyntaxKind["LessEqualOperator"] = 44] = "LessEqualOperator";
    // Syntax Sugar `x > y` <=> `not (y <= x)`
    // but we should probably use them to make the output code more readable
    SyntaxKind[SyntaxKind["GreaterThanOperator"] = 45] = "GreaterThanOperator";
    SyntaxKind[SyntaxKind["GreaterEqualOperator"] = 46] = "GreaterEqualOperator";
    // Logical
    SyntaxKind[SyntaxKind["AndOperator"] = 47] = "AndOperator";
    SyntaxKind[SyntaxKind["OrOperator"] = 48] = "OrOperator";
    SyntaxKind[SyntaxKind["NotOperator"] = 49] = "NotOperator";
    // Bitwise
    SyntaxKind[SyntaxKind["BitwiseAndOperator"] = 50] = "BitwiseAndOperator";
    SyntaxKind[SyntaxKind["BitwiseOrOperator"] = 51] = "BitwiseOrOperator";
    SyntaxKind[SyntaxKind["BitwiseExclusiveOrOperator"] = 52] = "BitwiseExclusiveOrOperator";
    SyntaxKind[SyntaxKind["BitwiseRightShiftOperator"] = 53] = "BitwiseRightShiftOperator";
    SyntaxKind[SyntaxKind["BitwiseLeftShiftOperator"] = 54] = "BitwiseLeftShiftOperator";
    SyntaxKind[SyntaxKind["BitwiseNotOperator"] = 55] = "BitwiseNotOperator";
})(SyntaxKind = exports.SyntaxKind || (exports.SyntaxKind = {}));
var NodeFlags;
(function (NodeFlags) {
    NodeFlags[NodeFlags["None"] = 0] = "None";
    NodeFlags[NodeFlags["Inline"] = 1] = "Inline";
    NodeFlags[NodeFlags["Declaration"] = 2] = "Declaration";
    NodeFlags[NodeFlags["TableUnpackCall"] = 4] = "TableUnpackCall";
})(NodeFlags = exports.NodeFlags || (exports.NodeFlags = {}));
function createNode(kind, tsOriginal) {
    if (tsOriginal === undefined) {
        return { kind, flags: NodeFlags.None };
    }
    const sourcePosition = getSourcePosition(tsOriginal);
    if (sourcePosition) {
        return { kind, line: sourcePosition.line, column: sourcePosition.column, flags: NodeFlags.None };
    }
    else {
        return { kind, flags: NodeFlags.None };
    }
}
exports.createNode = createNode;
function cloneNode(node) {
    return { ...node };
}
exports.cloneNode = cloneNode;
function setNodePosition(node, position) {
    node.line = position.line;
    node.column = position.column;
    return node;
}
exports.setNodePosition = setNodePosition;
function setNodeOriginal(node, tsOriginal) {
    if (node === undefined) {
        return undefined;
    }
    const sourcePosition = getSourcePosition(tsOriginal);
    if (sourcePosition) {
        setNodePosition(node, sourcePosition);
    }
    return node;
}
exports.setNodeOriginal = setNodeOriginal;
function getSourcePosition(sourceNode) {
    var _a;
    const parseTreeNode = (_a = ts.getParseTreeNode(sourceNode)) !== null && _a !== void 0 ? _a : sourceNode;
    const sourceFile = parseTreeNode.getSourceFile();
    if (sourceFile !== undefined && parseTreeNode.pos >= 0) {
        const { line, character } = ts.getLineAndCharacterOfPosition(sourceFile, parseTreeNode.pos + parseTreeNode.getLeadingTriviaWidth());
        return { line, column: character };
    }
}
function getOriginalPos(node) {
    return { line: node.line, column: node.column };
}
exports.getOriginalPos = getOriginalPos;
function setNodeFlags(node, flags) {
    node.flags = flags;
    return node;
}
exports.setNodeFlags = setNodeFlags;
function isFile(node) {
    return node.kind === SyntaxKind.File;
}
exports.isFile = isFile;
function createFile(statements, luaLibFeatures, trivia, tsOriginal) {
    const file = createNode(SyntaxKind.File, tsOriginal);
    file.statements = statements;
    file.luaLibFeatures = luaLibFeatures;
    file.trivia = trivia;
    return file;
}
exports.createFile = createFile;
function isBlock(node) {
    return node.kind === SyntaxKind.Block;
}
exports.isBlock = isBlock;
function createBlock(statements, tsOriginal) {
    const block = createNode(SyntaxKind.Block, tsOriginal);
    block.statements = statements;
    return block;
}
exports.createBlock = createBlock;
function isDoStatement(node) {
    return node.kind === SyntaxKind.DoStatement;
}
exports.isDoStatement = isDoStatement;
function createDoStatement(statements, tsOriginal) {
    const statement = createNode(SyntaxKind.DoStatement, tsOriginal);
    statement.statements = statements;
    return statement;
}
exports.createDoStatement = createDoStatement;
function isVariableDeclarationStatement(node) {
    return node.kind === SyntaxKind.VariableDeclarationStatement;
}
exports.isVariableDeclarationStatement = isVariableDeclarationStatement;
function createVariableDeclarationStatement(left, right, tsOriginal) {
    const statement = createNode(SyntaxKind.VariableDeclarationStatement, tsOriginal);
    statement.left = (0, utils_1.castArray)(left);
    if (right)
        statement.right = (0, utils_1.castArray)(right);
    return statement;
}
exports.createVariableDeclarationStatement = createVariableDeclarationStatement;
function isAssignmentStatement(node) {
    return node.kind === SyntaxKind.AssignmentStatement;
}
exports.isAssignmentStatement = isAssignmentStatement;
function createAssignmentStatement(left, right, tsOriginal) {
    const statement = createNode(SyntaxKind.AssignmentStatement, tsOriginal);
    statement.left = (0, utils_1.castArray)(left);
    statement.right = right ? (0, utils_1.castArray)(right) : [];
    return statement;
}
exports.createAssignmentStatement = createAssignmentStatement;
function isIfStatement(node) {
    return node.kind === SyntaxKind.IfStatement;
}
exports.isIfStatement = isIfStatement;
function createIfStatement(condition, ifBlock, elseBlock, tsOriginal) {
    const statement = createNode(SyntaxKind.IfStatement, tsOriginal);
    statement.condition = condition;
    statement.ifBlock = ifBlock;
    statement.elseBlock = elseBlock;
    return statement;
}
exports.createIfStatement = createIfStatement;
function isIterationStatement(node) {
    return (node.kind === SyntaxKind.WhileStatement ||
        node.kind === SyntaxKind.RepeatStatement ||
        node.kind === SyntaxKind.ForStatement ||
        node.kind === SyntaxKind.ForInStatement);
}
exports.isIterationStatement = isIterationStatement;
function isWhileStatement(node) {
    return node.kind === SyntaxKind.WhileStatement;
}
exports.isWhileStatement = isWhileStatement;
function createWhileStatement(body, condition, tsOriginal) {
    const statement = createNode(SyntaxKind.WhileStatement, tsOriginal);
    statement.body = body;
    statement.condition = condition;
    return statement;
}
exports.createWhileStatement = createWhileStatement;
function isRepeatStatement(node) {
    return node.kind === SyntaxKind.RepeatStatement;
}
exports.isRepeatStatement = isRepeatStatement;
function createRepeatStatement(body, condition, tsOriginal) {
    const statement = createNode(SyntaxKind.RepeatStatement, tsOriginal);
    statement.body = body;
    statement.condition = condition;
    return statement;
}
exports.createRepeatStatement = createRepeatStatement;
function isForStatement(node) {
    return node.kind === SyntaxKind.ForStatement;
}
exports.isForStatement = isForStatement;
function createForStatement(body, controlVariable, controlVariableInitializer, limitExpression, stepExpression, tsOriginal) {
    const statement = createNode(SyntaxKind.ForStatement, tsOriginal);
    statement.body = body;
    statement.controlVariable = controlVariable;
    statement.controlVariableInitializer = controlVariableInitializer;
    statement.limitExpression = limitExpression;
    statement.stepExpression = stepExpression;
    return statement;
}
exports.createForStatement = createForStatement;
function isForInStatement(node) {
    return node.kind === SyntaxKind.ForInStatement;
}
exports.isForInStatement = isForInStatement;
function createForInStatement(body, names, expressions, tsOriginal) {
    const statement = createNode(SyntaxKind.ForInStatement, tsOriginal);
    statement.body = body;
    statement.names = names;
    statement.expressions = expressions;
    return statement;
}
exports.createForInStatement = createForInStatement;
function isGotoStatement(node) {
    return node.kind === SyntaxKind.GotoStatement;
}
exports.isGotoStatement = isGotoStatement;
function createGotoStatement(label, tsOriginal) {
    const statement = createNode(SyntaxKind.GotoStatement, tsOriginal);
    statement.label = label;
    return statement;
}
exports.createGotoStatement = createGotoStatement;
function isLabelStatement(node) {
    return node.kind === SyntaxKind.LabelStatement;
}
exports.isLabelStatement = isLabelStatement;
function createLabelStatement(name, tsOriginal) {
    const statement = createNode(SyntaxKind.LabelStatement, tsOriginal);
    statement.name = name;
    return statement;
}
exports.createLabelStatement = createLabelStatement;
function isReturnStatement(node) {
    return node.kind === SyntaxKind.ReturnStatement;
}
exports.isReturnStatement = isReturnStatement;
function createReturnStatement(expressions, tsOriginal) {
    const statement = createNode(SyntaxKind.ReturnStatement, tsOriginal);
    statement.expressions = expressions;
    return statement;
}
exports.createReturnStatement = createReturnStatement;
function isBreakStatement(node) {
    return node.kind === SyntaxKind.BreakStatement;
}
exports.isBreakStatement = isBreakStatement;
function createBreakStatement(tsOriginal) {
    return createNode(SyntaxKind.BreakStatement, tsOriginal);
}
exports.createBreakStatement = createBreakStatement;
function isExpressionStatement(node) {
    return node.kind === SyntaxKind.ExpressionStatement;
}
exports.isExpressionStatement = isExpressionStatement;
function createExpressionStatement(expressions, tsOriginal) {
    const statement = createNode(SyntaxKind.ExpressionStatement, tsOriginal);
    statement.expression = expressions;
    return statement;
}
exports.createExpressionStatement = createExpressionStatement;
function isNilLiteral(node) {
    return node.kind === SyntaxKind.NilKeyword;
}
exports.isNilLiteral = isNilLiteral;
function createNilLiteral(tsOriginal) {
    return createNode(SyntaxKind.NilKeyword, tsOriginal);
}
exports.createNilLiteral = createNilLiteral;
function isBooleanLiteral(node) {
    return node.kind === SyntaxKind.TrueKeyword || node.kind === SyntaxKind.FalseKeyword;
}
exports.isBooleanLiteral = isBooleanLiteral;
function createBooleanLiteral(value, tsOriginal) {
    return createNode(value ? SyntaxKind.TrueKeyword : SyntaxKind.FalseKeyword, tsOriginal);
}
exports.createBooleanLiteral = createBooleanLiteral;
function isDotsLiteral(node) {
    return node.kind === SyntaxKind.DotsKeyword;
}
exports.isDotsLiteral = isDotsLiteral;
function createDotsLiteral(tsOriginal) {
    return createNode(SyntaxKind.DotsKeyword, tsOriginal);
}
exports.createDotsLiteral = createDotsLiteral;
function isArgLiteral(node) {
    return node.kind === SyntaxKind.ArgKeyword;
}
exports.isArgLiteral = isArgLiteral;
function createArgLiteral(tsOriginal) {
    return createNode(SyntaxKind.ArgKeyword, tsOriginal);
}
exports.createArgLiteral = createArgLiteral;
function isNumericLiteral(node) {
    return node.kind === SyntaxKind.NumericLiteral;
}
exports.isNumericLiteral = isNumericLiteral;
function createNumericLiteral(value, tsOriginal) {
    const expression = createNode(SyntaxKind.NumericLiteral, tsOriginal);
    expression.value = value;
    return expression;
}
exports.createNumericLiteral = createNumericLiteral;
function isStringLiteral(node) {
    return node.kind === SyntaxKind.StringLiteral;
}
exports.isStringLiteral = isStringLiteral;
function createStringLiteral(value, tsOriginal) {
    const expression = createNode(SyntaxKind.StringLiteral, tsOriginal);
    expression.value = value;
    return expression;
}
exports.createStringLiteral = createStringLiteral;
function isLiteral(node) {
    return (isNilLiteral(node) ||
        isDotsLiteral(node) ||
        isArgLiteral(node) ||
        isBooleanLiteral(node) ||
        isNumericLiteral(node) ||
        isStringLiteral(node));
}
exports.isLiteral = isLiteral;
function isFunctionExpression(node) {
    return node.kind === SyntaxKind.FunctionExpression;
}
exports.isFunctionExpression = isFunctionExpression;
function createFunctionExpression(body, params, dots, flags = NodeFlags.None, tsOriginal) {
    const expression = createNode(SyntaxKind.FunctionExpression, tsOriginal);
    expression.body = body;
    expression.params = params;
    expression.dots = dots;
    expression.flags = flags;
    return expression;
}
exports.createFunctionExpression = createFunctionExpression;
function isTableFieldExpression(node) {
    return node.kind === SyntaxKind.TableFieldExpression;
}
exports.isTableFieldExpression = isTableFieldExpression;
function createTableFieldExpression(value, key, tsOriginal) {
    const expression = createNode(SyntaxKind.TableFieldExpression, tsOriginal);
    expression.value = value;
    expression.key = key;
    return expression;
}
exports.createTableFieldExpression = createTableFieldExpression;
function isTableExpression(node) {
    return node.kind === SyntaxKind.TableExpression;
}
exports.isTableExpression = isTableExpression;
function createTableExpression(fields = [], tsOriginal) {
    const expression = createNode(SyntaxKind.TableExpression, tsOriginal);
    expression.fields = fields;
    return expression;
}
exports.createTableExpression = createTableExpression;
function isUnaryExpression(node) {
    return node.kind === SyntaxKind.UnaryExpression;
}
exports.isUnaryExpression = isUnaryExpression;
function createUnaryExpression(operand, operator, tsOriginal) {
    const expression = createNode(SyntaxKind.UnaryExpression, tsOriginal);
    expression.operand = operand;
    expression.operator = operator;
    return expression;
}
exports.createUnaryExpression = createUnaryExpression;
function isBinaryExpression(node) {
    return node.kind === SyntaxKind.BinaryExpression;
}
exports.isBinaryExpression = isBinaryExpression;
function createBinaryExpression(left, right, operator, tsOriginal) {
    const expression = createNode(SyntaxKind.BinaryExpression, tsOriginal);
    expression.left = left;
    expression.right = right;
    expression.operator = operator;
    return expression;
}
exports.createBinaryExpression = createBinaryExpression;
function isCallExpression(node) {
    return node.kind === SyntaxKind.CallExpression;
}
exports.isCallExpression = isCallExpression;
function createCallExpression(expression, params, tsOriginal) {
    const callExpression = createNode(SyntaxKind.CallExpression, tsOriginal);
    callExpression.expression = expression;
    callExpression.params = params;
    return callExpression;
}
exports.createCallExpression = createCallExpression;
function isMethodCallExpression(node) {
    return node.kind === SyntaxKind.MethodCallExpression;
}
exports.isMethodCallExpression = isMethodCallExpression;
function createMethodCallExpression(prefixExpression, name, params, tsOriginal) {
    const callExpression = createNode(SyntaxKind.MethodCallExpression, tsOriginal);
    callExpression.prefixExpression = prefixExpression;
    callExpression.name = name;
    callExpression.params = params;
    return callExpression;
}
exports.createMethodCallExpression = createMethodCallExpression;
function isIdentifier(node) {
    return node.kind === SyntaxKind.Identifier;
}
exports.isIdentifier = isIdentifier;
function createIdentifier(text, tsOriginal, symbolId, originalName) {
    const expression = createNode(SyntaxKind.Identifier, tsOriginal);
    expression.exportable = true;
    expression.text = text;
    expression.symbolId = symbolId;
    expression.originalName = originalName;
    return expression;
}
exports.createIdentifier = createIdentifier;
function cloneIdentifier(identifier, tsOriginal) {
    return createIdentifier(identifier.text, tsOriginal, identifier.symbolId, identifier.originalName);
}
exports.cloneIdentifier = cloneIdentifier;
function createAnonymousIdentifier(tsOriginal) {
    const expression = createNode(SyntaxKind.Identifier, tsOriginal);
    expression.exportable = false;
    expression.text = "____";
    return expression;
}
exports.createAnonymousIdentifier = createAnonymousIdentifier;
function isTableIndexExpression(node) {
    return node.kind === SyntaxKind.TableIndexExpression;
}
exports.isTableIndexExpression = isTableIndexExpression;
function createTableIndexExpression(table, index, tsOriginal) {
    const expression = createNode(SyntaxKind.TableIndexExpression, tsOriginal);
    expression.table = table;
    expression.index = index;
    return expression;
}
exports.createTableIndexExpression = createTableIndexExpression;
function isAssignmentLeftHandSideExpression(node) {
    return isIdentifier(node) || isTableIndexExpression(node);
}
exports.isAssignmentLeftHandSideExpression = isAssignmentLeftHandSideExpression;
function isFunctionDefinition(statement) {
    var _a;
    return statement.left.length === 1 && ((_a = statement.right) === null || _a === void 0 ? void 0 : _a.length) === 1 && isFunctionExpression(statement.right[0]);
}
exports.isFunctionDefinition = isFunctionDefinition;
function isInlineFunctionExpression(expression) {
    var _a;
    return (((_a = expression.body.statements) === null || _a === void 0 ? void 0 : _a.length) === 1 &&
        isReturnStatement(expression.body.statements[0]) &&
        expression.body.statements[0].expressions !== undefined &&
        (expression.flags & NodeFlags.Inline) !== 0);
}
exports.isInlineFunctionExpression = isInlineFunctionExpression;
